extmod/btstack/modbluetooth_btstack.c

```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+    MICROPY_PY_BLUETOOTH_EXIT
+}
+
+// Register a pending background operation -- copies the buffer, and makes it known to the GC.
+STATIC mp_btstack_pending_op_t* btstack_enqueue_pending_operation(uint16_t op_type, uint16_t conn_handle, uint16_t value_handle, const uint8_t *buf, size_t len) {
+    DEBUG_EVENT_printf("btstack_enqueue_pending_operation op_type=%d conn_handle=%d value_handle=%d len=%lu\n", op_type, conn_handle, value_handle, len);
+    mp_btstack_pending_op_t *pending_op = m_malloc(sizeof(mp_btstack_pending_op_t) + len);
+    pending_op->op_type = op_type;
+    pending_op->conn_handle = conn_handle;
+    pending_op->value_handle = value_handle;
+    pending_op->len = len;
+    memcpy(pending_op->buf, buf, len);
+
+    if (op_type == MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY || op_type == MP_BLUETOOTH_BTSTACK_PENDING_INDICATE) {
+        pending_op->context_registration.callback = &btstack_notify_indicate_ready_handler;
+        pending_op->context_registration.context = pending_op;
+    }
+
+    MICROPY_PY_BLUETOOTH_ENTER
+    assert(btstack_linked_list_add(&MP_STATE_PORT(bluetooth_btstack_root_pointers)->pending_ops, (btstack_linked_item_t*)pending_op));
+    MICROPY_PY_BLUETOOTH_EXIT
+
+    return pending_op;
+}
+
+#if MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
+// Cleans up a pending op of the specified type for this conn_handle (and if specified, value_handle).
+// Used by MP_BLUETOOTH_BTSTACK_PENDING_WRITE and MP_BLUETOOTH_BTSTACK_PENDING_WRITE_NO_RESPONSE.
+// At the moment, both will set value_handle=0xffff as the events do not know their value_handle.
+// TODO: Can we make btstack give us the value_handle for regular write (with response) so that we
+// know for sure that we're using the correct entry.
+STATIC mp_btstack_pending_op_t* btstack_finish_pending_operation(uint16_t op_type, uint16_t conn_handle, uint16_t value_handle) {
+    MICROPY_PY_BLUETOOTH_ENTER
+    DEBUG_EVENT_printf("btstack_finish_pending_operation op_type=%d conn_handle=%d value_handle=%d\n", op_type, conn_handle, value_handle);
+    btstack_linked_list_iterator_t it;
+    btstack_linked_list_iterator_init(&it, &MP_STATE_PORT(bluetooth_btstack_root_pointers)->pending_ops);
+    while (btstack_linked_list_iterator_has_next(&it)){
+        mp_btstack_pending_op_t *pending_op = (mp_btstack_pending_op_t*)btstack_linked_list_iterator_next(&it);
+
+        if (pending_op->op_type == op_type && pending_op->conn_handle == conn_handle && (value_handle == 0xffff || pending_op->value_handle == value_handle)) {
+            DEBUG_EVENT_printf("btstack_finish_pending_operation: found value_handle=%d len=%lu\n", pending_op->value_handle, pending_op->len);
+            assert(btstack_linked_list_remove(&MP_STATE_PORT(bluetooth_btstack_root_pointers)->pending_ops, (btstack_linked_item_t*)pending_op));
+            MICROPY_PY_BLUETOOTH_EXIT
+            return pending_op;
```

Comment:
as above, is it possible to explicitly free the `buf` before returning?

Reply:
The caller of this function needs it (e.g. `event_type == GATT_EVENT_CAN_WRITE_WITHOUT_RESPONSE` in the packet handler). But all the callers can free it.